odilia/src/tower.rs

```diff
@@ -0,0 +1,258 @@
+#![allow(dead_code)]
+
+use atspi::events::{document::DocumentEvents, object::ObjectEvents};
+use atspi::AtspiError;
+use atspi::Event;
+use atspi::GenericEvent;
+use odilia_common::errors::OdiliaError;
+use std::future::Future;
+use std::marker::PhantomData;
+
+use futures::future::{err, join_all, ok, Either, FutureExt, JoinAll, Ready};
+use futures::{Stream, StreamExt};
+use std::collections::HashMap;
+use std::task::Context;
+use std::task::Poll;
+
+use tower::util::BoxService;
+use tower::Layer;
+use tower::Service;
+
+#[derive(Debug)]
+pub enum Command {
+    Speak(String),
+}
+
+type Response = Vec<Command>;
+type Request = Event;
+type Error = OdiliaError;
+
+pub struct Handlers<S> {
+	state: S,
+	atspi_handlers: HashMap<(String, String), Vec<BoxService<Event, Response, Error>>>,
+  command_handlers: HashMap<Command, Vec<BoxService<Command, (), Error>>>,
+}
+impl<S> Handlers<S>
+where
+	S: Clone + Send + Sync + 'static,
+{
+	pub fn new(state: S) -> Self {
+		Handlers { state, atspi_handlers: HashMap::new(), command_handlers: HashMap::new() }
+	}
+	pub async fn atspi_handler<R>(mut self, mut events: R)
+	where
+		R: Stream<Item = Result<Event, AtspiError>> + Unpin,
+	{
+		std::pin::pin!(&mut events);
+		while let Some(Ok(ev)) = events.next().await {
+			let r = match ev {
+				Event::Object(ObjectEvents::StateChanged(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Object(ObjectEvents::TextCaretMoved(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Object(ObjectEvents::ChildrenChanged(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Object(ObjectEvents::TextChanged(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Document(DocumentEvents::LoadComplete(e)) => {
+					self.call_event_listeners(e).await
+				}
+				_ => {
+					println!("Not implemented yet....");
+					vec![Ok(vec![])]
+				}
+			};
+			for res in r {
+				if let Ok(resp) = res {
+				} else if let Err(err) = res {
+					println!("ERR: {:?}", err);
+				}
+			}
+		}
+	}
+	async fn call_event_listeners<'a, E>(&mut self, ev: E) -> Vec<Result<Response, Error>>
+	where
+		E: atspi::GenericEvent<'a> + Into<Event> + Send + Sync + 'a,
+	{
+		let dn = (
+			<E as atspi::GenericEvent<'a>>::DBUS_MEMBER.into(),
+			<E as atspi::GenericEvent<'a>>::DBUS_INTERFACE.into(),
+		);
+		let input = ev.into();
+    // NOTE: Why not use join_all(...) ?
+    // Because this drives the futures concurrently, and we want ordered handlers.
```

Comment:
I did mention in there that I can not have them compute concurrently, and why. The `Vec<Command>` may need to modify state.

Reply:
isn't this the purpose of this refactor, to remove state handling within handlers, but modify state externally outside the handler? if we modify global state in there, we're back to no testability or very hard testability

Reply:
A set of `Command`s *are instructions on how to modify the state*. Yes, the idea of this is to move the handling of state separate from the handling of actions. For example, an event may update speak some text, update the cache, fire a key event, or open a remote connection.

The implementations of these actions should be separate from their intention. If I want to speak x and y, then open a remote connection, that should be done through a set of commands, which are applied onto the state.

This way, we de-duplicate code, and de-couple the implmentation. So now, an addon can say "Hey, if I get this event, then I saw to beep at this frequency, or play this sound if some other condition is matched", and Odilia can respond without trying to give direct access to resources across FFI boundaries.

Reply:
but then, state still doesn't have to be updated in the handler, no? and then, event handlers can indeed run in paralell

Reply:
State will (eventually) be applied within a single call to a boxed service. But, since getting partial state is required (mostly) for each function, the functions that need state will need to be run in series to avoid sending out copies of stale data. Imagine that F1 updates the currently focused item, and F2 depends on the state of the currently focused item. If run in parallel, these functions will both get a copy of the current focused item at time 0, F1 will then run, producing a command to update the currently focused item, then F2 runs saying to speak the currently focused item and its role. This will result in speaking the *last focused item* instead of the current one.